### ISSUE TITLE
ETS leak from "typer"/"els_code_lens_suggest_spec" lense

### ISSUE THREAD

**Issue by seriyps** [erlang-ls/erlang_ls#952]
**Describe the bug**
Already discussed in private, but leaving it here for the reference.

It seems that `els_code_lens_suggest_spec` leaks ETS tables quite heavily. I noticed that, with this lense enabled, the number of ETS tables (as well as RAM allocated for ETS) in erlang_ls node starts to grow dramatically each time I do some changes, especially in a big file.

For example, here is my erlang_ls node after ~10minutes of me working on a large Erlang file. I have 4k+ open ETS tables, all belonging to `els_code_lens_provider` process:

![image](https://user-images.githubusercontent.com/422014/111922329-73102780-8a99-11eb-945e-6b66ec71fbba.png)
![image](https://user-images.githubusercontent.com/422014/111922334-79060880-8a99-11eb-98b7-baf296ddd92f.png)

**To Reproduce**
Enable `suggest-spec` lens, do a few changes in a file that has a lots of functions without typespecs. Observe that the number of ETS tables in erlang_ls node grows and not decreases after every edit (eg, via observer or `erlang:system_info(ets_count)`).

**Expected behavior**
Number of ETS tables stays constant.

**Actual behavior**
Number of ETS tables owned by `els_code_lens_provider` grows after each edit.

**Context**
 - `erlang_ls` version (tag/sha): 356a9ea28db91371f4cf144a6849cced04dd472d
 - Editor used: Emacs
 - LSP client used: lsp-mode

**Comment by seriyps** [comment]
Looking at the code of `els_typer`, it looks like `els_typer:extract` calls `CodeServer = dialyzer_codeserver:new()` which creates lots of ETS tables:

```erlang
7> rr(dialyzer_codeserver).               
[analysis,codeserver,contract,options]
8> CodeServer = dialyzer_codeserver:new().
#codeserver{next_core_label = 0,
            code = #Ref<0.2153276918.899809287.204834>,
            exported_types = #Ref<0.2153276918.899809287.204838>,
            records = #Ref<0.2153276918.899809287.204837>,
            contracts = #Ref<0.2153276918.899809287.204835>,
            callbacks = #Ref<0.2153276918.899809287.204836>,
            fun_meta_info = #Ref<0.2153276918.899809287.204840>,
            exports = #Ref<0.2153276918.899809287.204839>,
            temp_exported_types = #Ref<0.2153276918.899809287.204841>,
            temp_records = #Ref<0.2153276918.899809287.204842>,
            temp_contracts = #Ref<0.2153276918.899809287.204843>,
            temp_callbacks = #Ref<0.2153276918.899809287.204844>}
```

but never actually closes them.

So, there are 2 options:

* patch `els_typer` to close those tables (eg, by calling `dialyzer_code_server:delete(CodeServer)`)
* spawn a temporary process, so, ETS tables are cleaned when temporary process exits
* reuse those tables instead of re-creating them each time

**Comment by robertoaloi** [comment]
Should be fixed by #950 . Please re-open if that's not the case.

**Comment by seriyps** [comment]
Yep, I tried it and it indeed no longer leaks ETS tables :+1: 
Might be good to try to start caching / sharing those tables between lense executions in the future though. But it would be not a bugfix but a feature / optimization then.
